Updated codecommit process to reflect new EKS environment (#4069)
alphagov/govuk-developer-docs: source/manual/github-unavailable.html.md

---
owner_slack: "#govuk-developers"
title: When GitHub is unavailable
section: Deployment
layout: manual_layout
parent: "/manual.html"
---

## Backup to AWS CodeCommit

> We mirror all non-archived GitHub repositories tagged with `govuk` to AWS CodeCommit. This is achieved using a Mirror GitHub repositories workflow, which is [configured in terraform](https://github.com/alphagov/govuk-infrastructure/blob/main/terraform/deployments/github/mirror.tf).

Reviewer: We manage the code commit repositories (i.e. create/delete the repos) using Terraform. But the actual syncing the data between GitHub and those CodeCommit repos happens in a GitHub Action workflow here:https://github.com/alphagov/govuk-infrastructure/blob/main/.github/workflows/mirror-repos.yml

Author: Have updated to include reference to the Github action.

parent: "/manual.html"
---

## Backup to AWS CodeCommit

> We mirror all non-archived GitHub repositories tagged with `govuk` to AWS CodeCommit. This is achieved using a Mirror GitHub repositories workflow, which is [configured in terraform](https://github.com/alphagov/govuk-infrastructure/blob/main/terraform/deployments/github/mirror.tf) and carried out by this [github workflow](https://github.com/alphagov/govuk-infrastructure/blob/main/.github/workflows/mirror-repos.yml).

Reviewer 2: I think removing the repeated "workflow" bit will make this more succinct:

```suggestion
We mirror all non-archived GitHub repositories tagged with `govuk` to AWS CodeCommit. This is achieved using a ["Mirror repositories" GitHub workflow](https://github.com/alphagov/govuk-infrastructure/blob/main/.github/workflows/mirror-repos.yml), which uses a [`github_action_mirror_repos_role` IAM role configured in Terraform](https://github.com/alphagov/govuk-infrastructure/blob/main/terraform/deployments/github/mirror.tf).
```

### Checkout code in AWS CodeCommit

Since our deployment images are named via git tags, we'll need to checkout our code from AWS CodeCommit in order to make sure we're deploying the correct image and have an up-to-date copy of the code.

Setup your local environment for [checking out code from CodeCommit](/manual/howto-checkout-and-commit-to-codecommit.html)

## Drill creating and deploying a branch from CodeCommit

> If GitHub.com is down, old images can still be deployed manually in Argo.

Reviewer 2: I think the "old" reference is a throwback from before you'd figured out how to create and deploy new images?

```suggestion
If GitHub.com is down, Docker images can still be created and then deployed manually in Argo.
```

> 1. Follow the steps for [checking out and commiting from CodeCommit](/manual/howto-checkout-and-commit-to-codecommit.html) to make a change.

Reviewer 2: ```suggestion
### Create and deploy an image to ECR

1. Follow the steps for [checking out and commiting from CodeCommit](/manual/howto-checkout-and-commit-to-codecommit.html) to make a change.
```

> 1. Follow the steps for [checking out and commiting from CodeCommit](/manual/howto-checkout-and-commit-to-codecommit.html) to make a change.
> 1. Ensure the following instructions are all run from the repository root directory.

Reviewer 2: ```suggestion
Ensure the following instructions are all run from the root directory of the repository.

1. Follow the steps for [checking out and committing from CodeCommit](/manual/howto-checkout-and-commit-to-codecommit.html) to clone the repo.
1. Make your changes.
```

> 1. Get the IMAGE_TAG for your change as you will need to push this to the ECR manually in order to deploy the change from Argo.

Reviewer 2: First time we've mentioned ECR in the document, so worth expanding it:

```suggestion
1. Name the `IMAGE_TAG` for your change, which you will then push to the ECR (Elastic Container Registry) in order to deploy the change from Argo.
```

```
LOCAL_HEAD_SHA="$(git rev-parse HEAD)"
IMAGE_TAG="release-${LOCAL_HEAD_SHA}"
ECR_REGISTRY="172025368201.dkr.ecr.eu-west-1.amazonaws.com"
REPO=$(awk -F '=' /GOVUK_APP_NAME/'{print $2}' Dockerfile)
```

1. Build the docker image locally with the tag named after the latest sha.

```
 docker build --platform linux/amd64 -f Dockerfile -t $IMAGE_TAG .
```

> 1. Push the image to ECR after logging in.

Reviewer 2: Clearer wording, otherwise it sounds like you need to log in before running the step below:

```suggestion
1. Log in and push the image to ECR:
```

```
gds aws govuk-production-poweruser aws ecr get-login-password --region eu-west-1 | docker login --username AWS --password-stdin $ECR_REGISTRY
docker push $ECR_REGISTRY/$REPO:$IMAGE_TAG
```

> 1. Get the Admin Argo credentials by running `kubectl -n cluster-services get secret argocd-initial-admin-secret -oyaml |yq .data.password |base64 -d`

Reviewer 2: ```suggestion
### Deploy the image to EKS using Argo

1. Get the Admin Argo credentials by running `kubectl -n cluster-services get secret argocd-initial-admin-secret -oyaml |yq .data.password |base64 -d`
```

1. Log into [Argo](https://argo.eks.integration.govuk.digital/)
1. Search for the application you wish to deploy in the "Search applications..." search bar
1. Select the application from the list below the search bar
1. Select "APP DETAILS"
1. Go to "DISABLE AUTO-SYNC" then select "ok" when prompted "Disable Auto-Sync?"

> 1. Close the "APP DETAILS" menu now select the three dots next to the Deploy object for the part of the application you'd like to redeploy.

Reviewer 2: ```suggestion
1. Close the "APP DETAILS" menu, then select the three dots next to the Deploy object for the part of the application you'd like to redeploy. For example, `account-api-worker` would be for deploying the `account-api-workers`.
```

> 1. Go to the "LIVE MANIFEST" and select "EDIT". find the "image" then update it to the correct image you'd like to deploy. e.g account-api-worker would be for deploying the the account-api-workers.

Reviewer 2: ```suggestion
1. Go to the "LIVE MANIFEST" and select "EDIT". Find the "image", then update it to the new image tag you'd like to deploy. (You can double-check you have the correct image name by doing a `git log` after [checking out the code from CodeCommit](/manual/howto-checkout-and-commit-to-codecommit.html)).
```

> 1. You can double-check you have the correct image name by doing a `git log` after [checking out the code from CodeCommit](/manual/howto-checkout-and-commit-to-codecommit.html).

Reviewer 2: I've merged the above point into the one above it. I'd suggest finishing off with some summary, e.g.

```suggestion
1. Click "Save" and Argo will start the deployment, which should complete in under five minutes (??)
```

1. If you are prompted to add the item to keychain, deny.

There is more information about setting up your access key in the [AWS guide](https://docs.aws.amazon.com/codecommit/latest/userguide/setting-up-https-unixes.html)

> ## Drill creating and deploying a branch from CodeCommit

Reviewer 2: This is linked to from https://docs.publishing.service.gov.uk/manual/2nd-line-drills.html#deploy-from-aws-codecommit-when-github-is-unavailable

By deleting this section, are you saying it's no longer possible to drill deploying from CodeCommit?

Indeed, is it no longer possible to deploy anything when GitHub is down? ISTR an offline chat with Sean suggesting it could still be done but would require building images locally and pushing the images up to the store. Can we document that process instead?

Author: The tools for deploying from codecommit (git and jenkins) are being decommissioned so yes that drill is no longer possible. 
I'll speak with @theseanything and confirm if there is a way to deploy when github is down.

Author: I've updated the process with a way to deploy previously deployed versions of code that are already stored in ECR.
